## Data Cleaning

* Este notebook es la primera parte del proyecto de visualización para el bootcamp IronHack - semana 5
* Se trata de una base de datos descargada de kaggle del siguiente link: https://www.kaggle.com/code/marwandiab/boat-sales-analysis

In [26]:
import pandas as pd
#import string as str

import sys
sys.path.append('../')
import src.support as sp

In [2]:
data = pd.read_csv('../data/boat_data.csv')
data.head()

Price            Boat Type          Manufacturer                 Type  \
0   CHF 3337          Motor Yacht  Rigiflex power boats  new boat from stock   
1   EUR 3490  Center console boat     Terhi power boats  new boat from stock   
2   CHF 3770           Sport Boat    Marine power boats  new boat from stock   
3  DKK 25900           Sport Boat    Pioner power boats  new boat from stock   
4   EUR 3399         Fishing Boat    Linder power boats  new boat from stock   

   Year Built  Length  Width       Material  \
0        2017    4.00   1.90            NaN   
1        2020    4.00   1.50  Thermoplastic   
2           0    3.69   1.42      Aluminium   
3        2020    3.00   1.00            NaN   
4        2019    3.55   1.46      Aluminium   

                                     Location  Number of views last 7 days  
0      Switzerland » Lake Geneva » Vésenaz                          226  
1                    Germany » Bönningstedt                           75  
2  Switzerland » Lake of Zurich » Stäfa ZH                          124  
3                        Denmark » Svendborg                           64  
4               Germany » Bayern » München                           58

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9888 entries, 0 to 9887
Data columns (total 10 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Price                        9888 non-null   object 
 1   Boat Type                    9888 non-null   object 
 2   Manufacturer                 8550 non-null   object 
 3   Type                         9882 non-null   object 
 4   Year Built                   9888 non-null   int64  
 5   Length                       9879 non-null   float64
 6   Width                        9832 non-null   float64
 7   Material                     8139 non-null   object 
 8   Location                     9852 non-null   object 
 9   Number of views last 7 days  9888 non-null   int64  
dtypes: float64(2), int64(2), object(6)
memory usage: 772.6+ KB


Tenemos 9888 entradas, podemos ver en el .info() que hay datos nulos, sobre todo en la columna de 'Manufacturer' 

Se va a proceder a limpiar la muestra de valores nulos para poder trabajar con una visualización clara y limpia.

También se va a separar la columna 'Price' para separar en dos: valor numérico y moneda.

In [4]:
data.describe().T

count         mean         std    min      25%  \
Year Built                   9888.0  1893.192860  460.201582   0.00  1996.00   
Length                       9879.0    11.570017    6.002820   1.04     7.47   
Width                        9832.0     3.520124    1.220534   0.01     2.54   
Number of views last 7 days  9888.0   149.160801  151.819752  13.00    70.00   

                                 50%      75%      max  
Year Built                   2007.00  2017.00  2021.00  
Length                         10.28    13.93   100.00  
Width                           3.33     4.25    25.16  
Number of views last 7 days   108.00   172.00  3263.00

In [5]:
# Llamamos a la función y separamos en dos columnas aplicando función a toda la columna
data[['Precio', 'Currency']] = data['Price'].apply(sp.separa_valor_monedas).apply(pd.Series)

In [6]:
data.head()

Price            Boat Type          Manufacturer                 Type  \
0   CHF 3337          Motor Yacht  Rigiflex power boats  new boat from stock   
1   EUR 3490  Center console boat     Terhi power boats  new boat from stock   
2   CHF 3770           Sport Boat    Marine power boats  new boat from stock   
3  DKK 25900           Sport Boat    Pioner power boats  new boat from stock   
4   EUR 3399         Fishing Boat    Linder power boats  new boat from stock   

   Year Built  Length  Width       Material  \
0        2017    4.00   1.90            NaN   
1        2020    4.00   1.50  Thermoplastic   
2           0    3.69   1.42      Aluminium   
3        2020    3.00   1.00            NaN   
4        2019    3.55   1.46      Aluminium   

                                     Location  Number of views last 7 days  \
0      Switzerland » Lake Geneva » Vésenaz                          226   
1                    Germany » Bönningstedt                           75   
2  Switzerland » Lake of Zurich » Stäfa ZH                          124   
3                        Denmark » Svendborg                           64   
4               Germany » Bayern » München                           58   

    Precio Currency  
0   3337.0      CHF  
1   3490.0      EUR  
2   3770.0      CHF  
3  25900.0      DKK  
4   3399.0      EUR

In [7]:
data.Currency.unique()

array(['CHF', 'EUR', 'DKK', '£'], dtype=object)

In [8]:
data.Currency.value_counts()

Currency
EUR    8430
CHF     980
£      298
DKK     180
Name: count, dtype: int64

In [9]:
# Voy a limpiar todas aquellas donde aparezca £ y cambiarlo por GBP
data.Currency.replace('£','GBP', inplace=True)

In [10]:
data.Currency.value_counts()

Currency
EUR    8430
CHF     980
GBP     298
DKK     180
Name: count, dtype: int64

In [11]:
data['Precio_EUR'] = data.apply(sp.convert_currency, axis=1)
data.head()

Price            Boat Type          Manufacturer                 Type  \
0   CHF 3337          Motor Yacht  Rigiflex power boats  new boat from stock   
1   EUR 3490  Center console boat     Terhi power boats  new boat from stock   
2   CHF 3770           Sport Boat    Marine power boats  new boat from stock   
3  DKK 25900           Sport Boat    Pioner power boats  new boat from stock   
4   EUR 3399         Fishing Boat    Linder power boats  new boat from stock   

   Year Built  Length  Width       Material  \
0        2017    4.00   1.90            NaN   
1        2020    4.00   1.50  Thermoplastic   
2           0    3.69   1.42      Aluminium   
3        2020    3.00   1.00            NaN   
4        2019    3.55   1.46      Aluminium   

                                     Location  Number of views last 7 days  \
0      Switzerland » Lake Geneva » Vésenaz                          226   
1                    Germany » Bönningstedt                           75   
2  Switzerland » Lake of Zurich » Stäfa ZH                          124   
3                        Denmark » Svendborg                           64   
4               Germany » Bayern » München                           58   

    Precio Currency  Precio_EUR  
0   3337.0      CHF     3470.48  
1   3490.0      EUR     3490.00  
2   3770.0      CHF     3920.80  
3  25900.0      DKK     3367.00  
4   3399.0      EUR     3399.00

* Ya tengo los datos que necesito, puedo deshacerme de la columna 'Price' y poner en su lugar 'Precio', 'Currency' y 'Precio_EUR'

In [12]:
data = data.drop('Price', axis=1)

In [13]:
data.head()

Boat Type          Manufacturer                 Type  Year Built  \
0          Motor Yacht  Rigiflex power boats  new boat from stock        2017   
1  Center console boat     Terhi power boats  new boat from stock        2020   
2           Sport Boat    Marine power boats  new boat from stock           0   
3           Sport Boat    Pioner power boats  new boat from stock        2020   
4         Fishing Boat    Linder power boats  new boat from stock        2019   

   Length  Width       Material                                    Location  \
0    4.00   1.90            NaN      Switzerland » Lake Geneva » Vésenaz   
1    4.00   1.50  Thermoplastic                    Germany » Bönningstedt   
2    3.69   1.42      Aluminium  Switzerland » Lake of Zurich » Stäfa ZH   
3    3.00   1.00            NaN                        Denmark » Svendborg   
4    3.55   1.46      Aluminium               Germany » Bayern » München   

   Number of views last 7 days   Precio Currency  Precio_EUR  
0                          226   3337.0      CHF     3470.48  
1                           75   3490.0      EUR     3490.00  
2                          124   3770.0      CHF     3920.80  
3                           64  25900.0      DKK     3367.00  
4                           58   3399.0      EUR     3399.00

In [14]:
data.columns

Index(['Boat Type', 'Manufacturer', 'Type', 'Year Built', 'Length', 'Width',
       'Material', 'Location', 'Number of views last 7 days', 'Precio',
       'Currency', 'Precio_EUR'],
      dtype='object')

In [15]:
# Voy a reordenar las columnas como me apetece:
data = data[['Precio','Currency', 'Precio_EUR', 'Boat Type', 'Manufacturer', 'Type', 'Year Built', 'Length', 'Width',
       'Material', 'Location', 'Number of views last 7 days']]
data.head()

Precio Currency  Precio_EUR            Boat Type          Manufacturer  \
0   3337.0      CHF     3470.48          Motor Yacht  Rigiflex power boats   
1   3490.0      EUR     3490.00  Center console boat     Terhi power boats   
2   3770.0      CHF     3920.80           Sport Boat    Marine power boats   
3  25900.0      DKK     3367.00           Sport Boat    Pioner power boats   
4   3399.0      EUR     3399.00         Fishing Boat    Linder power boats   

                  Type  Year Built  Length  Width       Material  \
0  new boat from stock        2017    4.00   1.90            NaN   
1  new boat from stock        2020    4.00   1.50  Thermoplastic   
2  new boat from stock           0    3.69   1.42      Aluminium   
3  new boat from stock        2020    3.00   1.00            NaN   
4  new boat from stock        2019    3.55   1.46      Aluminium   

                                     Location  Number of views last 7 days  
0      Switzerland » Lake Geneva » Vésenaz                          226  
1                    Germany » Bönningstedt                           75  
2  Switzerland » Lake of Zurich » Stäfa ZH                          124  
3                        Denmark » Svendborg                           64  
4               Germany » Bayern » München                           58

In [16]:
# Ahora voy a cambiar los NaNs de la columna material con 'unknown'
data.Material.fillna('unknown', inplace=True)

In [17]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9888 entries, 0 to 9887
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Precio                       9888 non-null   float64
 1   Currency                     9888 non-null   object 
 2   Precio_EUR                   9888 non-null   float64
 3   Boat Type                    9888 non-null   object 
 4   Manufacturer                 8550 non-null   object 
 5   Type                         9882 non-null   object 
 6   Year Built                   9888 non-null   int64  
 7   Length                       9879 non-null   float64
 8   Width                        9832 non-null   float64
 9   Material                     9888 non-null   object 
 10  Location                     9852 non-null   object 
 11  Number of views last 7 days  9888 non-null   int64  
dtypes: float64(4), int64(2), object(6)
memory usage: 927.1+ KB


In [18]:
# Y procedo igual con manufacturer:
data.Manufacturer.fillna('unknown', inplace=True)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9888 entries, 0 to 9887
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Precio                       9888 non-null   float64
 1   Currency                     9888 non-null   object 
 2   Precio_EUR                   9888 non-null   float64
 3   Boat Type                    9888 non-null   object 
 4   Manufacturer                 9888 non-null   object 
 5   Type                         9882 non-null   object 
 6   Year Built                   9888 non-null   int64  
 7   Length                       9879 non-null   float64
 8   Width                        9832 non-null   float64
 9   Material                     9888 non-null   object 
 10  Location                     9852 non-null   object 
 11  Number of views last 7 days  9888 non-null   int64  
dtypes: float64(4), int64(2), object(6)
memory usage: 927.1+ KB


In [19]:
# El resto de valores nulos que haya los voy a borrar porque es info importante
data.dropna(inplace=True)
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9790 entries, 0 to 9887
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Precio                       9790 non-null   float64
 1   Currency                     9790 non-null   object 
 2   Precio_EUR                   9790 non-null   float64
 3   Boat Type                    9790 non-null   object 
 4   Manufacturer                 9790 non-null   object 
 5   Type                         9790 non-null   object 
 6   Year Built                   9790 non-null   int64  
 7   Length                       9790 non-null   float64
 8   Width                        9790 non-null   float64
 9   Material                     9790 non-null   object 
 10  Location                     9790 non-null   object 
 11  Number of views last 7 days  9790 non-null   int64  
dtypes: float64(4), int64(2), object(6)
memory usage: 994.3+ KB


In [20]:
data.head(10)

Precio Currency  Precio_EUR            Boat Type          Manufacturer  \
0    3337.0      CHF     3470.48          Motor Yacht  Rigiflex power boats   
1    3490.0      EUR     3490.00  Center console boat     Terhi power boats   
2    3770.0      CHF     3920.80           Sport Boat    Marine power boats   
3   25900.0      DKK     3367.00           Sport Boat    Pioner power boats   
4    3399.0      EUR     3399.00         Fishing Boat    Linder power boats   
5    3650.0      CHF     3796.00           Sport Boat    Linder power boats   
6    3600.0      CHF     3744.00            Catamaran               unknown   
8    3333.0      EUR     3333.00         Fishing Boat  Crescent power boats   
9    3300.0      EUR     3300.00         Pontoon Boat     Whaly power boats   
10   3500.0      CHF     3640.00         Fishing Boat     Terhi power boats   

                   Type  Year Built  Length  Width       Material  \
0   new boat from stock        2017    4.00   1.90        unknown   
1   new boat from stock        2020    4.00   1.50  Thermoplastic   
2   new boat from stock           0    3.69   1.42      Aluminium   
3   new boat from stock        2020    3.00   1.00        unknown   
4   new boat from stock        2019    3.55   1.46      Aluminium   
5   new boat from stock           0    4.03   1.56      Aluminium   
6    Used boat,Unleaded        1999    6.20   2.38      Aluminium   
8   new boat from stock        2019    3.64   1.37        unknown   
9   new boat from stock        2018    4.35   1.73        unknown   
10   Used boat,Electric        1987    4.35   1.75            GRP   

                                         Location  Number of views last 7 days  
0          Switzerland » Lake Geneva » Vésenaz                          226  
1                        Germany » Bönningstedt                           75  
2      Switzerland » Lake of Zurich » Stäfa ZH                          124  
3                            Denmark » Svendborg                           64  
4                   Germany » Bayern » München                           58  
5         Switzerland » Lake Constance » Uttwil                          132  
6       Switzerland » Neuenburgersee » Yvonand                           474  
8   Germany » Bayern » Boote+service Oberbayern                           45  
9                           Italy » Dormelletto                           180  
10                         Switzerland » Seengen                          239

Ahora voy a pasar a limpiar la columna 'Location' porque parece hay varios caracteres feos
* Lo voy a hacer de manera local y no con una función externa, solo aplica a este caso.

In [21]:
data[['Country', 'City']] = data.Location.str.split('»', n=1, expand=True) # Separo por '»' por verlo en el head() como separa País y ciudades

data[['City', 'Other']] = data.City.str.split('»', n=1, expand=True) # Separo la nueva 'City' en otros -> a veces hay estado o mas detalles de localización

data.City = data.City.str.replace('©', '', regex=True) # Limpio también caracteres raros
data.City = data.City.str.replace('¶', '', regex=True)
data.City = data.City.str.replace('Ã', '', regex=True)

data.head(10) # A ver como ha quedado


Precio Currency  Precio_EUR            Boat Type          Manufacturer  \
0    3337.0      CHF     3470.48          Motor Yacht  Rigiflex power boats   
1    3490.0      EUR     3490.00  Center console boat     Terhi power boats   
2    3770.0      CHF     3920.80           Sport Boat    Marine power boats   
3   25900.0      DKK     3367.00           Sport Boat    Pioner power boats   
4    3399.0      EUR     3399.00         Fishing Boat    Linder power boats   
5    3650.0      CHF     3796.00           Sport Boat    Linder power boats   
6    3600.0      CHF     3744.00            Catamaran               unknown   
8    3333.0      EUR     3333.00         Fishing Boat  Crescent power boats   
9    3300.0      EUR     3300.00         Pontoon Boat     Whaly power boats   
10   3500.0      CHF     3640.00         Fishing Boat     Terhi power boats   

                   Type  Year Built  Length  Width       Material  \
0   new boat from stock        2017    4.00   1.90        unknown   
1   new boat from stock        2020    4.00   1.50  Thermoplastic   
2   new boat from stock           0    3.69   1.42      Aluminium   
3   new boat from stock        2020    3.00   1.00        unknown   
4   new boat from stock        2019    3.55   1.46      Aluminium   
5   new boat from stock           0    4.03   1.56      Aluminium   
6    Used boat,Unleaded        1999    6.20   2.38      Aluminium   
8   new boat from stock        2019    3.64   1.37        unknown   
9   new boat from stock        2018    4.35   1.73        unknown   
10   Used boat,Electric        1987    4.35   1.75            GRP   

                                         Location  \
0          Switzerland » Lake Geneva » Vésenaz   
1                        Germany » Bönningstedt   
2      Switzerland » Lake of Zurich » Stäfa ZH   
3                            Denmark » Svendborg   
4                   Germany » Bayern » München   
5         Switzerland » Lake Constance » Uttwil   
6       Switzerland » Neuenburgersee » Yvonand    
8   Germany » Bayern » Boote+service Oberbayern   
9                           Italy » Dormelletto    
10                         Switzerland » Seengen   

    Number of views last 7 days       Country              City  \
0                           226  Switzerland       Lake Geneva    
1                            75      Germany        Bnningstedt   
2                           124  Switzerland    Lake of Zurich    
3                            64      Denmark          Svendborg   
4                            58      Germany            Bayern    
5                           132  Switzerland    Lake Constance    
6                           474  Switzerland    Neuenburgersee    
8                            45      Germany            Bayern    
9                           180        Italy       Dormelletto    
10                          239  Switzerland            Seengen   

                        Other  
0                    Vésenaz  
1                        None  
2                   Stäfa ZH  
3                        None  
4                    München  
5                      Uttwil  
6                    Yvonand   
8    Boote+service Oberbayern  
9                        None  
10                       None

In [22]:
# Una vez separado en tres columnas, a simple vista puedo ver que la columna 'Other' no va a aportar gra información y puede traer problemas 
# así que voy a deshacerme de ella, así como quitar la columna Location y ordenar a mi gusto.

data = data.drop(['Location', 'Other'], axis=1)

In [23]:
data.columns

Index(['Precio', 'Currency', 'Precio_EUR', 'Boat Type', 'Manufacturer', 'Type',
       'Year Built', 'Length', 'Width', 'Material',
       'Number of views last 7 days', 'Country', 'City'],
      dtype='object')

In [24]:
data = data[['Precio', 'Currency', 'Precio_EUR', 'Boat Type', 'Manufacturer', 'Type',
       'Year Built', 'Length', 'Width', 'Material', 'Country', 'City', 'Number of views last 7 days']]
data.head()

Precio Currency  Precio_EUR            Boat Type          Manufacturer  \
0   3337.0      CHF     3470.48          Motor Yacht  Rigiflex power boats   
1   3490.0      EUR     3490.00  Center console boat     Terhi power boats   
2   3770.0      CHF     3920.80           Sport Boat    Marine power boats   
3  25900.0      DKK     3367.00           Sport Boat    Pioner power boats   
4   3399.0      EUR     3399.00         Fishing Boat    Linder power boats   

                  Type  Year Built  Length  Width       Material  \
0  new boat from stock        2017    4.00   1.90        unknown   
1  new boat from stock        2020    4.00   1.50  Thermoplastic   
2  new boat from stock           0    3.69   1.42      Aluminium   
3  new boat from stock        2020    3.00   1.00        unknown   
4  new boat from stock        2019    3.55   1.46      Aluminium   

        Country              City  Number of views last 7 days  
0  Switzerland       Lake Geneva                           226  
1      Germany        Bnningstedt                           75  
2  Switzerland    Lake of Zurich                           124  
3      Denmark          Svendborg                           64  
4      Germany            Bayern                            58

De cara a tener una base de datos 'friendly' con PowerBI a cambiar los separadores de decimales de '.' a ',' para las oclumnas Precio_EUR, Length y Width

In [31]:
import pandas as pd

# Assuming 'Precio_EUR', 'Length', and 'Width' are columns in your DataFrame
data['Precio_EUR'] = data['Precio_EUR'].astype(str).apply(pd.Series.str.replace, pat='.', repl=',')
data['Length'] = data['Length'].astype(str).apply(pd.Series.str.replace, pat='.', repl=',')
data['Width'] = data['Width'].astype(str).apply(pd.Series.str.replace, pat='.', repl=',')

data.head()

data.head()

TypeError: Cannot interpret '<module 'string' from 'c:\\Users\\jacom\\anaconda3\\envs\\clase\\lib\\string.py'>' as a data type

In [34]:
# Listo, me quedo con esta base de datos. La paso a mi carpeta de data como boat_data_clean
data.to_csv('../data/boat_data_clean.csv')